From d0aeb8890da8d649d412af40faa7d4478d5295c7 Mon Sep 17 00:00:00 2001
From: Subbaraya Sundeep <sbhatta@marvell.com>
Date: Wed, 17 Feb 2021 14:16:34 +0530
Subject: [PATCH 12/12] octeontx2-af: Fix VF mac configuration by VF.

commit 9ddd4e87cb8c0bdfd6f4d022898463ce2e20a7b9 from
git@git.assembla.com:cavium/WindRiver.linux.git

PF can set its VF mac address using the
"ip link set dev <PF_eth_interface> vf <VF_id> mac <mac_addr>".
The above command changes the default unicast filter
installed by AF for the VF. Since PF is changing
the AF default unicast rule the command fails because
owner to mcam entry mapping fails. Fix this by relaxing
checks in AF incase PF sets mac for its VF.

Fixes: 878432cb ("octeontx2-af: refactor function npc_install_flow")
Change-Id: I62cb7138c3a692460ed33220ca7ff7306a8e8910
Signed-off-by: Subbaraya Sundeep <sbhatta@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/46179
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
index fe00a8863bbb..7449a1dc8c67 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_npc_fs.c
@@ -1047,6 +1047,13 @@ static int npc_install_flow(struct rvu *rvu, int blkaddr, u16 target,
 		rvu_mcam_remove_counter_from_rule(rvu, owner, rule);
 
 	write_req.hdr.pcifunc = owner;
+
+	/* AF owns the default rules so change the owner just to relax
+	 * the checks in rvu_mbox_handler_npc_mcam_write_entry
+	 */
+	if (req->default_rule)
+		write_req.hdr.pcifunc = 0;
+
 	write_req.entry = entry_index;
 	write_req.intf = req->intf;
 	write_req.enable_entry = (u8)enable;
-- 
2.26.1

